src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp

```diff
@@ -149,8 +154,14 @@ double ElkanKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
     // At this point, we know the new cluster assignment.
     // Step 4: for each center c, let m(c) be the mean of the points assigned to
     // c.
-    newCentroids.col(assignments[i]) += arma::vec(dataset.col(i));
-    counts[assignments[i]]++;
+    #pragma omp critical
```

Comment:
The critical sections are probably the pain points that prevent better speedup here.  I think that if we want to OpenMP-ize more than just distance calculations, we may need to hold sets of centroids for each individual core, and then merge them all together at the end.  (Writing that in a way that is agnostic to whether or not OpenMP is available may involve ifdefs... maybe even as simple as:

```
#ifdef _OPENMP // not sure this is the right macro, be sure to double-check
const size_t threads = omp_get_num_threads();
#else
const size_t threads = 1;
#endif
```

and then base the rest of your code around `threads` or something like this.  Just one of many possible ideas.)